BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -0,0 +1,433 @@
+    LocalCAN_.Init.TransmitFifoPriority = DISABLE;
+
+    /* initialize can perph */
+    if (HAL_CAN_Init(&LocalCAN_) != HAL_OK)
+    {
+        Error_Handler();
+    }
+
+    /* enable interrupts */
+    HAL_NVIC_EnableIRQ(CAN3_RX0_IRQn);
+    HAL_NVIC_EnableIRQ(CAN3_TX_IRQn);
+
+    /* start can perph */
+    HAL_CAN_Start(&LocalCAN_);
+}
+
+// void HAL_CAN_MspInit(CAN_HandleTypeDef *canHandle)
+// {
+//     GPIO_InitTypeDef GPIO_InitStruct = (GPIO_InitTypeDef){0};
+
+//     if (canHandle->Instance == CAN1)
+//     {
+
+//         /* CAN1 clock enable */
+//         __HAL_RCC_CAN1_CLK_ENABLE();
+
+//         /* GPIOA clock enable */
+//         __HAL_RCC_GPIOA_CLK_ENABLE();
+
+//         /**CAN1 GPIO Configuration
+//         PA11     ------> CAN1_RX
+//         PA12     ------> CAN1_TX
+//         */
+//         GPIO_InitStruct.Pin = GPIO_PIN_11 | GPIO_PIN_12;
+//         GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+//         GPIO_InitStruct.Pull = GPIO_NOPULL;
+//         GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+//         GPIO_InitStruct.Alternate = GPIO_AF9_CAN1;
+//         HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
+//     }
+//     else if (canHandle->Instance == CAN3)
+//     {
+//         /* CAN3 clock enable */
+//         __HAL_RCC_CAN3_CLK_ENABLE();
+
+//         /* GPIOA clock enable */
+//         __HAL_RCC_GPIOA_CLK_ENABLE();
+
+//         /**CAN3 GPIO Configuration
+//         PA8     ------> CAN3_RX
+//         PA15     ------> CAN3_TX
+//         */
+//         GPIO_InitStruct.Pin = GPIO_PIN_8 | GPIO_PIN_15;
+//         GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+//         GPIO_InitStruct.Pull = GPIO_NOPULL;
+//         GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+//         GPIO_InitStruct.Alternate = GPIO_AF11_CAN3;
+//         HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
+//     }
+// }
+
+void HAL_CAN_MspDeInit(CAN_HandleTypeDef *canHandle)
+{
+    if (canHandle->Instance == CAN1)
+    {
+        /* CAN1 clock disable */
+        __HAL_RCC_CAN1_CLK_DISABLE();
+
+        /**CAN1 GPIO Configuration
+        PA11     ------> CAN1_RX
+        PA12     ------> CAN1_TX
+        */
+        HAL_GPIO_DeInit(GPIOA, GPIO_PIN_11 | GPIO_PIN_12);
+    }
+    else if (canHandle->Instance == CAN3)
+    {
+        /* CAN3 clock disable */
+        __HAL_RCC_CAN3_CLK_DISABLE();
+
+        /**CAN3 GPIO Configuration
+        PA8     ------> CAN3_RX
+        PA15     ------> CAN3_TX
+        */
+        HAL_GPIO_DeInit(GPIOA, GPIO_PIN_8 | GPIO_PIN_15);
+    }
+}
+
+bool CAN_Send(const CAN_HandleTypeDef *can, CAN_TxPayload_t payload, bool blocking)
+{
+    /* TODO: error check payload */
+
+    /* determine which queue/semaphore to use */
+    QueueHandle_t queue = can == CarCAN ? CarCAN_Queue : LocalCAN_Queue;
+    SemaphoreHandle_t semaphore = can == CarCAN ? CarCAN_SemaphoreHandle : LocalCAN_SemaphoreHandle;
+
+    /* take access of the can bus */
+    if (blocking)
+    {
+        while(xSemaphoreTake(semaphore, portMAX_DELAY)) {}
```

Comment:
xSemaphoreTake returns false if the time to wait expires without the semaphore becoming available, and true when the semaphore is actually acquired. I'm not sure this should be a while loop rather than an if statement for the blocking case, since the function will hang until either the semaphore is acquired or the time runs out.

```suggestion
        if(!xSemaphoreTake(semaphore, portMAX_DELAY)) return false;
```